### PR TITLE
Use the Travis cache infrastructure and Pytest's xdist to cut test runtime in half.

### PR DESCRIPTION
**Pull request by kmaglione**

### CODE REVIEW

**Comment by kmaglione** [comment]
The test run in this repo won't show the full speed-up, since it doesn't have wheel caches yet. But it it still significantly faster than previously. A run in my repo, with populated caches, is about 3 minutes faster:

https://travis-ci.org/kmaglione/olympia/builds/83042288

**Comment by andymckay-limited-access** [comment]
Nice one!

File: conftest.py [hunk 1]
```diff
@@ -36,8 +48,6 @@ def prefix_indexes(config):
     else:
         prefix = 'test'
 
-    from django.conf import settings
```
Comment (magopian) [inline conftest.py]: where does the `settings` come from?

File: requirements/git.txt [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+#
+# 1) Update the portion of the git URL immediately following the `@` to the
+#    target commit hash.
+#
+# 2) Update the requirement line immediately following the git URL to point to
+#    the version number that commit corresponds to.
+#
+# 3) Build a new wheel for the updated git URL using `pip wheel <git-url>`.
+#
+# 4) Upload that wheel to our private pyrepo, or if you don't have access,
```
Comment (magopian) [inline requirements/git.txt]: Do you want to add a note explaining that the wheels should go in the `wheelhouse/` folder on pyrepo?

**Comment by magopian** [comment]
Not sure aboug `scripts/tox_install.sh` which looks brittle, could we instead make use of the new feature of recent pip that does automatic caching/wheel building?

File: conftest.py (hunk 1 above)
Comment (kmaglione) [inline conftest.py]: It's already being imported at the top of the file.

File: requirements/git.txt (hunk 2 above)
Comment (kmaglione) [inline requirements/git.txt]: Perhaps. We haven't been very consistent about it, though. There are wheels in both the wheelhouse folder and the parent pyrepo folder.

**Comment by kmaglione** [comment]
I'm not a huge fan of it either. It's the main reason I didn't submit this earlier. There are a few reasons just using the default pip caching doesn't work:

1) The pip HTTP cache is very bad. In theory, it should fetch our indexes once, and then read them from the cache for each package it needs to install. Or, really, it should download them once, compile them into an index, and then use that to find packages. What it actually does is fetch both the pyrepo and wheelhouse index once for each package in our requirements, which adds a minute to a minute and a half to the build time.

2) The git dependencies will get built for every installation, even if we already have a cached wheel. I could remove the automatic git builds from the `tox_install` script, which would simplify things a bit, and require temporary hacks or wheelhouse uploads for each test run that changed them. I thought about doing that, but ideally I'd like Travis to just deploy the built packages to our wheelhouse whenever we merge to master (once we have peep, anyway), so it seemed like it might be worth keeping.

That said, the script is much less fragile than it looks. I don't really like that it has to re-run the download job after it builds git wheels, but aside from that it should be pretty solid. `pip install --download` uses the same logic as a plain `pip install`, so when it succeeds, any package required by `pip install` is assured to be in the cache.

In the worst case, if the script breaks, we can just revert `.travis.yml` to do a plain install from our pyrepo and live with the perf hit for a while.

**Comment by andymckay** [comment]
This is such a nice perf win, that I feel that "perfect is the enemy of good" here. We can try it out, see what breaks and improve as we go.

**Comment by magopian** [comment]
Yep, ship it!

**Pull request merged**